libraries/persistent_store/src/driver.rs

```diff
@@ -0,0 +1,630 @@
+        model: usize,
+    },
+
+    /// The store failed to track the number of erase cycles correctly.
+    DifferentErase {
+        /// The first page in physical storage order with a wrong value.
+        page: usize,
+
+        /// How many times the page has been erased according to the store.
+        store: usize,
+
+        /// How many times the page has been erased according to the model.
+        model: usize,
+    },
+
+    /// The store failed to track the number of word writes correctly.
+    DifferentWrite {
+        /// The first page in physical storage order with a wrong value.
+        page: usize,
+
+        /// The first word in the page with a wrong value.
+        word: usize,
+
+        /// How many times the word has been written according to the store.
+        ///
+        /// This value is exact only for the metadata of the page. For the content of the page, it
+        /// is set to:
+        /// - 0 if the word is after the tail. Such word should not have been written.
+        /// - 1 if the word is before the tail. Such word may or may not have been written.
+        store: usize,
+
+        /// How many times the word has been written according to the model.
+        ///
+        /// This value is exact only for the metadata of the page. For the content of the page, it
+        /// is set to:
+        /// - 0 if the word was not written.
+        /// - 1 if the word was written.
+        model: usize,
+    },
+}
+
+impl StoreDriver {
+    /// Provides read-only access to the storage.
+    pub fn storage(&self) -> &BufferStorage {
+        match self {
+            StoreDriver::On(x) => x.store().storage(),
+            StoreDriver::Off(x) => x.storage(),
+        }
+    }
+
+    /// Extracts the power-on version of the driver.
+    pub fn on(self) -> Option<StoreDriverOn> {
+        match self {
+            StoreDriver::On(x) => Some(x),
+            StoreDriver::Off(_) => None,
+        }
+    }
+
+    /// Powers on the store if not already on.
+    pub fn power_on(self) -> Result<StoreDriverOn, StoreInvariant> {
+        match self {
+            StoreDriver::On(x) => Ok(x),
+            StoreDriver::Off(x) => x.power_on(),
+        }
+    }
+
+    /// Extracts the power-off version of the driver.
+    pub fn off(self) -> Option<StoreDriverOff> {
+        match self {
+            StoreDriver::On(_) => None,
+            StoreDriver::Off(x) => Some(x),
+        }
+    }
+}
+
+impl StoreDriverOff {
+    /// Starts a simulation with a clean storage given its configuration.
+    pub fn new(options: BufferOptions, num_pages: usize) -> StoreDriverOff {
+        let storage = vec![0xff; num_pages * options.page_size].into_boxed_slice();
+        let storage = BufferStorage::new(storage, options);
+        StoreDriverOff::new_dirty(storage)
+    }
+
+    /// Starts a simulation from an existing storage.
+    pub fn new_dirty(storage: BufferStorage) -> StoreDriverOff {
```

Comment:
Why is this public? (Similar for some other functions below.) I assume it's a high-control interface for fuzzing?

Reply:
Yes, this is used by fuzzing to start from an invalid state (simulating a user that did not clear the storage before installing a fresh OpenSK). This whole module is meant for fuzzing, but I also use it for tests because it's convenient and checks more invariants.